[Lee Jia Sheng] iP

jiasheng59 opened the pull request:
# Duke
> “Simplicity boils down to two steps: Identify the essential. Eliminate the rest.” - Leo Babauta ([source](https://www.goodreads.com/quotes/345851-simplicity-boils-down-to-two-steps-identify-the-essential-eliminate))
Duke will be your **personal** assistant to keep track of your tasks, thus allowing you to focus on those things that really matter to you. It features:
- Easy to learn
- Clear interface
- Free to use

Follow the steps below to get started:
1. Install it from this [website](https://github.com/nus-cs2103-AY2223S2/ip/pull/255)
2. Click _Run_
3. Now you can add your tasks! 😺 

Example use:
- [X] Go gym
- [x] Meeting
- [x] Watch movie

---

If you are interested in contributing to the development of Duke, check out our repository. Below is the code snippet of `main` method.
```Java
public static void main(String[] args) {
    new Duke().run();
}
```

Commit:
Add Gradle support

jiasheng59 committed:
Implement greet, echo and exit function

jiasheng59 committed:
Add a list to store tasks

jiasheng59 committed:
Add Task.java

jiasheng59 committed:
Implement mark task as done feature

jiasheng59 committed:
Change String representation of Task

jiasheng59 committed:
Add three subclasses of Task

jiasheng59 committed:
Add todo, deadline and event commands

jiasheng59 committed:
Testing text UI

jiasheng59 committed:
Implement DukeException.java

jiasheng59 committed:
Implement delete feature

jiasheng59 committed:
Add TaskType.java

jiasheng59 committed:
Implement save feature

jiasheng59 committed:
Merge branch 'branch-Level-7'

jiasheng59 committed:
Use LocalDateTime class to represent the date

jiasheng59 committed:
Resolve merge conflicts

jiasheng59 committed:
Merge branch-Level-8 onto master

jiasheng59 committed:
Refactor Duke class to make it more OOP

jiasheng59 committed:
Put all classes into package duke

jiasheng59 committed:
Merge branch 'add-gradle-support' of https://github.com/jiasheng59/ip

jiasheng59 committed:
Add test cases for Test and Parser classes

jiasheng59 committed:
Reorganize the file structure

jiasheng59 committed:
Add JavaDoc

jiasheng59 committed:
Merge branch 'branch-A-JavaDoc'

jiasheng59 committed:
Remove whitespaces

jiasheng59 committed:
Merge branch 'branch-A-CodingStandard'

jiasheng59 committed:
Implement find feature

jiasheng59 committed:
Merge branch 'branch-Level-9'

BoAi01 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private void run() {
+        Ui.printLogo();
```
I am not absolutely sure, but it might be better if there is a Javadoc for the method? :grinning:

BoAi01 commented on src/main/java/seedu/duke/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public String taskInFileFormat() {
+        return super.taskInFileFormat() + " | " + by;
+    }
```
If I remember correctly, it is better to use nouns for things (classes) and verbs for methods, so probably `getTaskInFileFormat` sounds better?

BoAi01 commented on src/main/java/seedu/duke/Parser.java:
```diff
@@ -0,0 +1,65 @@
+    private boolean isValidTask(String command) {
+        if (command.length() >= 4 && command.substring(0, 4).equals("todo")) {
+            if (command.length() <= 5) {
+                throw new DukeException("The description of a todo cannot be empty.");
```
I would personally prefer abstracting out checking whether a string is a certain command, since this is used in many places. Additionally, perhaps we want to avoid magic numbers like `4`. Replacing that with `"todo".length()` might improve readability.

BoAi01 commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                task = new Event(s[2], Ui.parseDateTime(date[0]), Ui.parseDateTime(date[1]));
+                break;
+            default:
+                throw new DukeException("Invalid type of task");
```
Nice indentation style!

BoAi01 commented on src/main/java/seedu/duke/Todo.java:
```diff
@@ -0,0 +1,13 @@
+package seedu.duke;
+
+public class Todo extends Task {
+    public Todo(String description) {
```
Perhaps it looks better if we have a Javadoc before the class declaration here?

BoAi01 commented on src/test/java/seedu/duke/TaskTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
```
If I recall correctly, `import static` statements should go before other `import` statements. Please correct me if I am wrong. :grinning:

BoAi01 commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+        int lineNumber = Math.toIntExact(Files.lines(filePath).count());
+        TaskList tasks = new TaskList(new ArrayList<>());
+        for (int i = 0; i < lineNumber; i++) {
+            String[] s = readFromDataFile(i);
```
Maybe this is not a big issue, but a potential improvement could be using a more descriptive name, such as `fileString`, instead of `s`.

BoAi01 reviewed (commented):
Overall, this shows good efforts! Most classes and methods are well documented, but there are still some Javadoc missing. Besides, it might be better if we use more descriptive names for variables in some places, and pay attention to the naming of methods (use verbs for class methods, as suggested [here](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html)). Some "magic numbers" can also be replaced with constants to improve readability. :smiley:

jiasheng59 committed:
Add basic GUI features

jiasheng59 committed:
Add checkstyle configuration files

jiasheng59 committed:
build.gradle: reorder plugins

jiasheng59 committed:
Merge branch 'branch-A-CheckStyle'

jiasheng59 committed:
build.gradle: add dependencies of Java Fx

jiasheng59 committed:
Ui class: modify print method

jiasheng59 committed:
Merge branch 'branch-A-Varargs'

tricixg commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,156 @@
+        stage.setScene(scene);
+        stage.show();
+
+        stage.setTitle("Duke");
```
Would be nice to have comments on what "stage" shows or contains

tricixg commented on src/main/java/seedu/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+     * @param start The start time of the event.
+     * @param end The end time of the event.
+     */
+    public Event(String description, LocalDateTime start, LocalDateTime end) {
```
javadocs well crafted

tricixg commented on src/main/java/seedu/duke/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        tasks.remove(index);
+    }
+
+    public void mark(int taskNumber, Storage storage) {
```
can add javadocs if there is time

tricixg reviewed (commented):
LGTM overall neat code and easy to understand

jiasheng59 committed:
Use assertions to handle invalid input

The entire code makes assumption that no invalid input is provided.

We want to handle the case where user entering invalid input like
zero or negative task number, searching for empty string, etc.

Add assert statements to ensure above mentioned cases are handled.

Assertions are more suitable way of handling this kind of edge cases
than using exceptions.

jiasheng59 committed:
Merge pull request #1 from jiasheng59/branch-A-Assertions

Use assertions to handle invalid input

jiasheng59 committed:
Storage class: refector fileToArrayList method

fileToArrayList contains different level of handling input logic in
single method.

This makes the code less clean and less readable.

Extract out the logic of creating a Task object from string to simply
the method body.

Follow the SLAP to ensure the method is easier to read and understand.

jiasheng59 committed:
Merge pull request #2 from jiasheng59/branch-A-CodeQuality

Storage class: refector fileToArrayList method

jiasheng59 committed:
Implement command to detect duplicate task

There's no way to check duplicate.

It's nice to have check_duplicate command as user may accidentally
add the same task twice.

Add check_duplicate command.

jiasheng59 committed:
Merge pull request #3 from jiasheng59/branch-C-DetectDuplicates

Implement command to detect duplicate task

jiasheng59 committed:
Fix delete command bug

jiasheng59 committed:
Ui class: extract out read command method

jiasheng59 committed:
Extract command into classes

jiasheng59 committed:
Extract command as classes

jiasheng59 committed:
Separate Gui from the main logic of duke

The gui is mixed with duke's logic of handling input in the Duke class.

This way of implementation makes it harder to further extend the Gui,
and makes the overall logic much more difficult to follow.

Extract the Gui portion in different classes, i.e. DialogBox,
MainWindow.

This is to allow developer to see the overview of code structure in
an easier way.

jiasheng59 committed:
Fixing code style

jiasheng59 committed:
Add a junit test for TaskList

jiasheng59 committed:
Remove unused code

jiasheng59 committed:
Complete all JavaDoc comments

jiasheng59 committed:
Finalise checkstyle

jiasheng59 committed:
Set up CI

jiasheng59 committed:
Remove I/O redirection test CI

jiasheng59 committed:
Add Ui.png

jiasheng59 committed:
Parser class: Fix parsing bug of todo

jiasheng59 committed:
Add README.md

jiasheng59 committed:
Fix mark done bug

jiasheng59 committed:
Fix trivial checkstyle for CI

jiasheng59 committed:
Fix Gui text overrun problem

jiasheng59 committed:
Update README.md

jiasheng59 committed:
Upload a new Ui.png

jiasheng59 committed:
Make gui resizable
